### PR TITLE
Connect: Replace legacy useStore hook with useStoreSelector in WorkspacesService

### PR DESCRIPTION
**Pull request by ravicious**
## The problem

During code review, I found that `DocumentCluster` with unified resources throws an error when logging out.

https://github.com/user-attachments/assets/67f69fb7-7497-4851-84e9-c0a5d1f299e0

`WorkspacesService` service holds state related to workspaces (AKA profiles), i.e. its root cluster, what tabs are open, etc. When the user logs out of a cluster, Connect removes state related to that workspace. The errors indicate that `DocumentCluster` was trying to read from `WorkspaceService` after the state related to the workspace was removed.

This caught my attention because it's not something that should happen. `DocumentsRenderer` reads from `WorkspacesService` first and then it renders components for each tab. When a workspace is removed, `DocumentsRenderer` no longer sees it, so the tab components for that workspace are unmounted – they shouldn't receive updates after the workspace gets removed.

After a chat with Grzegorz, we established that `DocumentCluster` must be somehow getting updates of `WorkspacesService` state faster than `DocumentsRenderer`. `DocumentsRenderer` was using only the legacy `useStore` hook, while `DocumentCluster` was in some places using the new `useStoreSelector`.

The hypothesis we had was that a component that uses `useStore` receives updates later than a component that uses `useStoreSelector`. This is because `useStore` uses a `useEffect` + `useState` combo while `useStoreSelector` uses  [`useSyncExternalStore`](https://react.dev/reference/react/useSyncExternalStore).

When the store gets updated, `useStore` needs at least three "ticks" - one to process the `onChange` callback submitted to `store.subscribe`, one to process `setState` (disguised as `rerender` function), and one to actually re-render the component. `useStoreSelector` needs only two ticks – one to process the subscription callback and one to re-render the component.

We're not sure how accurate this theory is, but a quick test of modifying `WorkspacesService.useState` to use `useStoreSelector` and `useStore` showed that this fixes the problem.

## React 19 and other improvements

When working on React 19, I found that the React Compiler linter was complaining a lot about hooks such as `WorkspacesService.useState`. They're defined as methods on a class which technically means that their identity can't be guaranteed. To satisfy the linter in the future, I decided to remove `useState` completely and replace it with `useWorkspacesServiceState`.

`useStoreSelector` can be much more performant than `useStore`, because it re-renders only when the selected state changes rather than on every change, even to an unrelated part of the state. In a few places, I was able to replace `useStore` with `useStoreSelector` and select a specific part of the state. Unfortunately, a lot of callsites depend on methods from `WorkspacesService` and `DocumentsService`. In those cases, I simply used `useWorkspacesServiceState`.

## Demo of the fix

https://github.com/user-attachments/assets/4be1137c-6430-4a6c-b26f-ff65d7681be1

### CODE REVIEW

File: web/packages/teleterm/src/ui/Search/SearchBar.tsx [hunk 1]
```diff
@@ -31,14 +31,17 @@ import {
...
 
 const OPEN_SEARCH_BAR_SHORTCUT_ACTION: KeyboardShortcutAction = 'openSearchBar';
 
 export function SearchBarConnected() {
-  const { workspacesService } = useAppContext();
-  workspacesService.useState();
+  const rootClusterUri = useStoreSelector(
+    'workspacesService',
+    useCallback(state => state.rootClusterUri, [])
+  );
```
Comment (avatus) [inline web/packages/teleterm/src/ui/Search/SearchBar.tsx:42]: i just had flashbacks of redux

**Review by avatus — APPROVED** [review]
Seems to work fine for me

**Review by gzdunek — APPROVED** [review]
I quickly tested the app, I didn't see any issues.

🚀

**Comment by ravicious** [comment]
This PR might have caused the access request sidebar to not progress to the next state on completion, I'll investigate that today.

**Pull request merged**